### PR TITLE
Feat/add snapshot based testing for helm

### LINKED ISSUE #1377: Add support for snapshot testing
**Describe the solution you'd like**
[helm-unittest](https://github.com/helm-unittest/helm-unittest) as a great feature called [snapshot testing](https://github.com/helm-unittest/helm-unittest#snapshot-testing).

Sometimes you may just want to check that the rendered manifest did not change between helm package version without every details asserted.

**Describe alternatives you've considered**
None and neither did the community, see https://github.com/orgs/gruntwork-io/discussions/780

**Additional context**

### PR DESCRIPTION
**Pull request by jguionnet**
## Description

Fixes #1377 

Add a set of helper functions, like the `UnmarshalK8SYaml` functions.

- `UpdateSnapshot` creates/refreshes a local snapshot
- `DiffAgainstSnapshot` diffs the manifest generated by the test against the local snapshot. It returns the number of differences that can be asserted in a test. It also displays the difference between the two sets of k8s manifests in git style.

See test output in this gist https://gist.github.com/jguionnet/376e768009a244d3315f31b82598aa78

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).
 
- [ x] Update the docs. (ie provided sample tests as documentation)
- [ x] Run the relevant tests successfully, including pre-commit checks.
- [ x ] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ x ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

Added support in Helm for snapshot testing, similar to [helm-unittest](https://github.com/helm-unittest/helm-unittest) feature [snapshot testing](https://github.com/helm-unittest/helm-unittest#snapshot-testing).  

### Migration Guide

NA

### CODE REVIEW

**Comment by denis256** [comment]
```
This branch has conflicts that must be resolved
Use the [web editor](https://github.com/gruntwork-io/terratest/pull/1378/conflicts) or the  to resolve conflicts.
Conflicting files
go.mod
```

**Comment by jguionnet** [comment]
@denis256: I resolved the conflict, could you please review it?

File: modules/helm/template_test.go [hunk 1]
```diff
@@ -65,3 +65,79 @@ func TestRemoteChartRender(t *testing.T) {
...
+	// we want to assert that the template renders without any errors.
+	output := RenderRemoteTemplate(t, options, remoteChartSource, releaseName, []string{})
+
+	// Now we use kubernetes/client-go library to render the template output into the Deployment struct. This will
+	// ensure the Deployment resource is rendered correctly.
+	var deployment appsv1.Deployment
+	UnmarshalK8SYaml(t, output, &deployment)
+
+	// run the diff and assert there is only one difference: the image name
+	number_of_diffs := DiffAgainstSnapshot(output, releaseName)
```
Comment (denis256) [inline modules/helm/template_test.go:141]: camelCase is preferred against snake_case, 
In function code is already defined snake case variables like `releaseName`

File: modules/helm/template.go [hunk 2]
```diff
@@ -134,3 +140,96 @@ func UnmarshalK8SYamlE(t testing.TestingT, yamlData string, destinationObj inter
 	}
 	return nil
 }
+
+// Add snapshot based testing for helm
+// see https://github.com/gruntwork-io/terratest/issues/1377
+// This function create/update the manifest snapshot of a chart (e.g bitnami/nginx)
+func UpdateSnapshot(yamlData string, releaseName string) {
```
[the remark this answers is not in the document]
Comment (jguionnet) [reply modules/helm/template.go:147]: Updated the function comment

**Comment by jguionnet** [comment]
@denis256. Thanks for reviewing. I fixed it according to the feedback.

File: modules/helm/template.go [hunk 3]
```diff
@@ -134,3 +140,99 @@ func UnmarshalK8SYamlE(t testing.TestingT, yamlData string, destinationObj inter
...
+	// write any difference to stdout
+	reportWriter := &dyff.HumanReport{
+		Report:            report,
+		DoNotInspectCerts: false,
+		NoTableStyle:      false,
+		OmitHeader:        false,
+		UseGoPatchPaths:   false,
+	}
+
+	reportWriter.WriteReport(os.Stdout)
```
Comment (denis256) [inline modules/helm/template.go:234]: Returned error from `reportWriter.WriteReport()` not handled

File: modules/helm/template.go [hunk 4]
```diff
@@ -134,3 +140,99 @@ func UnmarshalK8SYamlE(t testing.TestingT, yamlData string, destinationObj inter
...
+	fmt.Println("k8s manifest written into file: ", filename)
+}
+
+// DiffAgainstSnapshot compare the current manifests of a chart (e.g bitnami/nginx)
+// with the previous manifests stored in the snapshot.
+// see https://github.com/gruntwork-io/terratest/issues/1377
+// It returns the number of difference between the two manifest snaphost or -1 in case of error
+func DiffAgainstSnapshot(yamlData string, releaseName string) int {
+
+	snapshotDir := "__snapshot__"
```
Comment (denis256) [inline modules/helm/template.go:184]: why can't be used `helm.Options` struct to set snapshotDir ?

File: modules/helm/template.go [hunk 5]
```diff
@@ -134,3 +140,99 @@ func UnmarshalK8SYamlE(t testing.TestingT, yamlData string, destinationObj inter
...
+
+	// load the yaml snapshot file
+	snapshot := snapshotDir + "/" + releaseName + ".yaml"
+	from, err := ytbx.LoadFile(snapshot)
+	if err != nil {
+		fmt.Println("Error opening file:", err)
```
Comment (denis256) [inline modules/helm/template.go:190]: `logger.Logf()`

**Review by denis256 — COMMENTED** [review]
Proposed functions don't follow existing patterns:
- no `*E` function which check for errors
- `t testing.TestingT` not passed 

Go mod don't have all changes commited
```
$ git status
On branch feat/add-remote-repo-support-in-helm
nothing to commit, working tree clean
$ go mod tidy
$ git status
On branch feat/add-remote-repo-support-in-helm
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   go.mod


```

File: modules/helm/template.go [hunk 6]
```diff
@@ -134,3 +139,123 @@ func UnmarshalK8SYamlE(t testing.TestingT, yamlData string, destinationObj inter
...
+	if options.Logger != nil {
+		options.Logger.Logf(t, "helm chart manifest written into file: %s", filename)
+	}
+	return nil
+}
+
+// DiffAgainstSnapshot compare the current manifests of a chart (e.g bitnami/nginx)
+// with the previous manifests stored in the snapshot.
+// see https://github.com/gruntwork-io/terratest/issues/1377
+// It returns the number of difference between the two manifest snaphost or -1 in case of error
```
Comment (denis256) [inline modules/helm/template.go:193]: snapshot

File: modules/helm/template.go [hunk 7]
```diff
@@ -134,3 +139,123 @@ func UnmarshalK8SYamlE(t testing.TestingT, yamlData string, destinationObj inter
...
+// It will failed the test if there is an error while reading or writing the two manifests in the file system
```
Comment (denis256) [inline modules/helm/template.go:194]: fail

**Review by denis256 — COMMENTED** [review]
Approach looks fine, some wording updates and comments on functions should be updated

**Comment by jguionnet** [comment]
@denis256. Thank you for your thorough review and for catching all my mistakes. I fixed it according to your latest feedback.

**Review by denis256 — APPROVED** [review]

**Pull request merged**